proto/api/v1/auth_service.proto

```diff
@@ -20,21 +21,146 @@ message UserAttribute {
-        storage.ServiceIdentity       service_id  = 2;
+        string                  user_id         = 1;
+        storage.ServiceIdentity service_id      = 2;
     }
-    google.protobuf.Timestamp expires     = 3;
-    string                    refresh_url = 4;
+    google.protobuf.Timestamp   expires         = 3;
+    string                      refresh_url     = 4;
+    storage.AuthProvider        auth_provider   = 5;
+    storage.UserInfo            user_info       = 6;
+    repeated UserAttribute      user_attributes = 7;
+}
+
+// AuthMachineToMachineConfig determines rules for exchanging an identity token from a third party with
+// a Central access token. The M2M stands for machine to machine, as this is the intended use-case
+// for the config.
+message AuthMachineToMachineConfig {
+    // UUID of the config.
+    string id = 1;
+
+    // The type of the auth machine to machine config.
+    // Currently supports GitHub actions or any other generic OIDC provider to use for verifying and
+    // exchanging the token.
+    enum Type {
+        GITHUB_ACTIONS = 0;
+        GENERIC        = 1;
+    }
+    Type type          = 2;
+
+    // The expiration of the token annotated as a duration.
+    // Possible valid time units are: s, m, h.
+    // The maximum allowed expiration duration is 24h.
+    // As an example: 2h45m.
+    string token_expiration_duration = 3;
+
+   // Mappings map an identity token's claim values to a specific role within Central.
+    message Mapping {
+        // A key within the identity token's claim value to use.
+        string key            = 1;
+        // The value that should match to map the identity token to a specific role.
+        // This supports regular expressions of RE2 format, see more here:
+        // https://github.com/google/re2/wiki/Syntax.
+        string value          = 2;
+        // The role which should be issued when the key and value match for a particular identity token.
+        string role           = 3;
+    }
+    repeated Mapping mappings = 4;
+
+    // The issuer configuration for a generic OIDC issuer of the identity token.
+    // For the GitHub Action type, no additional configuration is required.
+    message GenericIssuer {
+        string issuer         = 1;
+    }
+    oneof IssuerConfig {
+        GenericIssuer generic = 5;
+    }
+}
+
+message ListAuthMachineToMachineConfigResponse {
+    repeated AuthMachineToMachineConfig configs = 1;
+}
+
+message GetAuthMachineToMachineConfigResponse {
+    AuthMachineToMachineConfig config = 1;
+}
+
+message AddAuthMachineToMachineConfigRequest {
+    AuthMachineToMachineConfig config = 1;
+}
+
+message AddAuthMachineToMachineConfigResponse {
+    AuthMachineToMachineConfig config = 1;
+}
+
+message UpdateAuthMachineToMachineConfigRequest {
+    AuthMachineToMachineConfig config = 1;
+}
+
+message UpdateAuthMachineToMachineConfigResponse {
+    AuthMachineToMachineConfig config = 1;
+}
```

Comment:
Yeah, so with `Add` only the updated ID will be of value, the rest _should_ be the same.
With `Update`, since `PUT` does allow also creating resources ([see RFC](https://datatracker.ietf.org/doc/html/rfc7231#section-4.3.4)), for completeness I'd favor returning it. Also, in case we ever would allow partial updates, it'd be better to return the complete object.

In general, this is mostly a personal preference, I do like when the complete object is returned compared to only e.g. the ID / nothing. There are a couple of API examples (like the role service) where things like the ID are not being returned. So, in theory I could minimize this to only returning the ID if wanted.

Reply:
Regarding `Add` - I'm ok with returning the complete object, not just the `id`.

Regarding `Update`:
- I'm still not certain that it's necessary to return back the same object. Haven't spotted anything in 4.3.4 about response body in the RFC.
- It is not in RFC semantics to support _partial_ updates with `PUT` method. `PUT` must receive the complete representation of the object in the input.
- Since `PUT` implies upsert semantics, the description of `UpdateAuthMachineToMachineConfig` must be updated. I'll create a separate comment for that.
- Also, the RFC suggests certain HTTP status codes. I'll create a separate comment for that.